packages/firestore/src/model/field_index.ts

```diff
@@ -60,11 +106,119 @@ export const enum Kind {
 }
 
 /** An index component consisting of field path and index type.  */
-export class Segment {
+export class IndexSegment {
   constructor(
     /** The field path of the component. */
     readonly fieldPath: FieldPath,
     /** The fields sorting order. */
-    readonly kind: Kind
+    readonly kind: IndexKind
+  ) {}
+}
+
+function indexSegmentComparator(
+  left: IndexSegment,
+  right: IndexSegment
+): number {
+  const cmp = FieldPath.comparator(left.fieldPath, right.fieldPath);
+  if (cmp !== 0) {
+    return cmp;
+  }
+  return primitiveComparator(left.kind, right.kind);
+}
+
+/** Stores the "high water mark" that indicates how updated the Index is for the current user. */
+export class IndexState {
+  constructor(
+    /** Indicates when the index was last updated (relative to other indexes). */
+    readonly sequenceNumber: number,
+    /** The the latest indexed read time, document and batch id. */
+    readonly offset: IndexOffset
   ) {}
+
+  /** The state of an index that has not yet been backfilled. */
+  static empty(): IndexState {
+    return new IndexState(INITIAL_SEQUENCE_NUMBER, IndexOffset.min());
+  }
+}
+
+/**
+ * Creates an offset that matches all documents with a read time higher than
+ * `readTime`.
+ */
+export function newIndexOffsetSuccessorFromReadTime(
+  readTime: SnapshotVersion,
+  largestBatchId: number
+): IndexOffset {
+  // We want to create an offset that matches all documents with a read time
+  // greater than the provided read time. To do so, we technically need to
+  // create an offset for `(readTime, MAX_DOCUMENT_KEY)`. While we could use
+  // Unicode codepoints to generate MAX_DOCUMENT_KEY, it is much easier to use
+  // `(readTime + 1, DocumentKey.empty())` since `> DocumentKey.empty()` matches
+  // all valid document IDs.
+  const successorSeconds = readTime.toTimestamp().seconds;
+  const successorNanos = readTime.toTimestamp().nanoseconds + 1;
+  const successor = SnapshotVersion.fromTimestamp(
+    successorNanos === 1e9
+      ? new Timestamp(successorSeconds + 1, 0)
+      : new Timestamp(successorSeconds, successorNanos)
+  );
+  return new IndexOffset(successor, DocumentKey.empty(), largestBatchId);
+}
+
+/** Creates a new offset based on the provided document. */
+export function newIndexOffsetFromDocument(document: Document): IndexOffset {
```

Comment:
Just double checking that you plan on adding another method to create a new index offset with a specified largest batch id.

Reply:
The equivalent for Web is just calling the constructor of IndexOffset.